### PR TITLE
Saving to tempo one by one

### LINKED ISSUE #80: Fix history of worklogs after emptying local storage hack

### PR DESCRIPTION
**Pull request by mikx98**
Closes #80, #101, #113, #102

### CODE REVIEW

File: src/actions/worklogActions.js [hunk 1]
```diff
@@ -124,14 +179,21 @@ export const moveCursor = (increment: number) => (
...
-  const { cursor, points } = getState()
+  const state = getState()
+  const { cursor } = state
+  const points = editablePointsSelector(state)
 
-  const [newCursor] = [cursor + increment]
+  const [newEditableIndex] = [cursor.editableIndex + increment]
     .map(acc => (acc === -1 ? points.length - 1 : acc))
     .map(acc => (acc === points.length ? 0 : acc))
     .map(acc => Math.max(acc, 0))
```
Comment (Siegrift) [inline src/actions/worklogActions.js]: You can use `clamp`.

File: src/worklogs/calendar/calendarSelectors.js [hunk 2]
```diff
@@ -74,7 +74,8 @@ export const calendarEventsSelector: InputSelector<
             : new Date(points[i + 1].timestamp),
         allDay: i === points.length - 1,
         cursor: isNormalMode && i === cursor,
-        selected: isNormalMode && selection.includes(i)
+        selected: isNormalMode && selection.includes(i),
+        synced: point.synced
```
[the remark this answers is not in the document]
Comment (mikx98) [reply src/worklogs/calendar/calendarSelectors.js]: maybe later, but i like this more and its also shorter :smile:

File: src/actions/worklogActions.js [hunk 3]
```diff
@@ -50,15 +64,21 @@ const editTime = (indices: number[], time: string) => ({
...
-    sortBy(
-      state.map((point, i) =>
-        payload.indices.includes(i)
-          ? { ...point, timestamp: payload.time }
-          : point
-      ),
-      ['timestamp']
+  reducer: (points, payload) => {
+    const invalidOperation = indices.some(
+      index => index > 0 && points[index - 1].synced
```
Comment (jakuub) [inline src/actions/worklogActions.js]: I think, we should check also if the point edited is not synced, not only previous point.

File: src/actions/worklogActions.js [hunk 4]
```diff
@@ -112,6 +137,36 @@ export const saveRecentIssue = (issue: string) => ({
   reducer: (state: string[]): string[] => [issue, ...state]
 })
 
+export const setPointsAsSynced = (syncedIds: number[]) => ({
+  type: 'Set points as synced',
+  payload: syncedIds,
+  path: ['points'],
+  reducer: (state: Point[]): Point[] =>
+    state.map((p, i) => ({ ...p, synced: p.synced || syncedIds.includes(i) }))
```
Comment (jakuub) [inline src/actions/worklogActions.js]: We expect here, that order/indices were not changed during request - is it somehow guaranteed?

File: src/worklogs/WorkLogs.js [hunk 5]
```diff
@@ -141,18 +141,15 @@ export default compose(
           },
           withCtrlKey: {
             j: () => openDialog(DIALOG_TYPES.JOIN_INTERVALS),
-            s: () => openDialog(DIALOG_TYPES.SUBMIT_WORKLOGS),
+            s: () => {
+              roundPoints()
```
[the remark this answers is not in the document]
Comment (mikx98) [reply src/worklogs/WorkLogs.js]: after discussion with @jakuub, for now i will just add `cancel and revert rounding` option next to submit, so people won't be desperate

in future, your idea for realtime rounding is possible, but not doable today

**Pull request merged**